src/ray/util/container_util.h

```diff
@@ -28,56 +30,92 @@
+  std::ostream &StringifyContainer(std::ostream &os) const {
+    os << "[";
+    for (auto it = obj_.begin(); it != obj_.end(); ++it) {
+      if (it != obj_.begin()) {
+        os << ", ";
+      }
+      os << debug_string(*it);
+    }
+    os << "]";
+    return os;
+  }
+
+  // Public but OK, since it's const &.
+  const T &obj_;
+};
+
+template <typename T>
+std::ostream &operator<<(std::ostream &os, DebugStringWrapper<T> wrapper) {
+  return os << wrapper.obj_;
 }
 
 template <typename... Ts>
-std::string debug_string(const std::pair<Ts...> &pair) {
-  std::stringstream ss;
-  ss << "(" << debug_string(pair.first) << ", " << debug_string(pair.second) << ")";
-  return ss.str();
+std::ostream &operator<<(std::ostream &os, DebugStringWrapper<std::pair<Ts...>> pair) {
+  return os << "(" << debug_string(pair.obj_.first) << ", "
+            << debug_string(pair.obj_.second) << ")";
 }
 
-template <typename C>
-std::string _container_debug_string(const C &c) {
-  std::stringstream ss;
-  ss << "[";
-  for (auto it = c.begin(); it != c.end(); ++it) {
-    if (it != c.begin()) {
-      ss << ", ";
-    }
-    ss << debug_string(*it);
+template <typename... Ts>
+std::ostream &operator<<(std::ostream &os, DebugStringWrapper<std::tuple<Ts...>> tuple) {
+  os << "(";
+  // This specialization is needed, or the compiler complains the lambda in std::apply
+  // does not use capture &os.
+  if constexpr (sizeof...(Ts) != 0) {
+    std::apply(
+        [&os](const Ts &...args) {
+          size_t n = 0;
+          ((os << debug_string(args) << (++n != sizeof...(Ts) ? ", " : "")), ...);
+        },
+        tuple.obj_);
   }
-  ss << "]";
-  return ss.str();
+  os << ")";
+  return os;
 }
 
 template <typename... Ts>
-std::string debug_string(const std::vector<Ts...> &c) {
-  return _container_debug_string(c);
+std::ostream &operator<<(std::ostream &os, DebugStringWrapper<std::vector<Ts...>> c) {
+  return c.StringifyContainer(os);
 }
 template <typename... Ts>
-std::string debug_string(const std::set<Ts...> &c) {
-  return _container_debug_string(c);
+std::ostream &operator<<(std::ostream &os, DebugStringWrapper<std::set<Ts...>> c) {
+  return c.StringifyContainer(os);
 }
 template <typename... Ts>
-std::string debug_string(const std::unordered_set<Ts...> &c) {
-  return _container_debug_string(c);
+std::ostream &operator<<(std::ostream &os,
+                         DebugStringWrapper<std::unordered_set<Ts...>> c) {
+  return c.StringifyContainer(os);
 }
 template <typename... Ts>
-std::string debug_string(const absl::flat_hash_set<Ts...> &c) {
-  return _container_debug_string(c);
+std::ostream &operator<<(std::ostream &os,
+                         DebugStringWrapper<absl::flat_hash_set<Ts...>> c) {
+  return c.StringifyContainer(os);
 }
 template <typename... Ts>
-std::string debug_string(const std::map<Ts...> &c) {
-  return _container_debug_string(c);
+std::ostream &operator<<(std::ostream &os, DebugStringWrapper<std::map<Ts...>> c) {
```

Comment:
I think using known type directly is a good starting point, but it would be nice to templatize.
For example, how do you plan to handle array-list container, like vector, array, inline vector, etc.

Reply:
Try checkout my implementation: https://github.com/dentiny/bitcask/tree/main/utils/strings

Reply:
Your implementation is really good. Though this `debug_string` is intended to be non-deserializable, very limited and we specifically only support several "blessed containers" to avoid accidental leak of user info (e.g. we can't print protobuf messages, though we can write good templates to support them).